src/drivers/roboclaw/RoboClaw.cpp

```diff
@@ -226,154 +312,207 @@ float RoboClaw::getMotorSpeed(e_motor motor)
+int RoboClaw::_sendNothing(e_command command)
 {
+	uint8_t recv_buff;
+	return _transaction(command, nullptr, 0, &recv_buff, 1);
+}
+
+uint16_t RoboClaw::_calcCRC(const uint8_t *buf, size_t n, uint16_t init)
+{
+	uint16_t crc = init;
+
+	for (size_t byte = 0; byte < n; byte++) {
+		crc = crc ^ (((uint16_t) buf[byte]) << 8);
+
+		for (uint8_t bit = 0; bit < 8; bit++) {
+			if (crc & 0x8000) {
+				crc = (crc << 1) ^ 0x1021;
+
+			} else {
+				crc = crc << 1;
+			}
+		}
+	}
+
+	return crc;
+}
+
+int RoboClaw::_transaction(e_command cmd, uint8_t *wbuff, size_t wbytes,
+			   uint8_t *rbuff, size_t rbytes, bool send_checksum, bool recv_checksum)
+{
+	int err_code = 0;
+
+	// WRITE
+
 	tcflush(_uart, TCIOFLUSH); // flush  buffers
-	uint8_t buf[n_data + 3];
+	uint8_t buf[wbytes + 4];
 	buf[0] = _address;
 	buf[1] = cmd;
 
-	for (size_t i = 0; i < n_data; i++) {
-		buf[i + 2] = data[n_data - i - 1]; // MSB
+	if (wbuff) {
+		memcpy(&buf[2], wbuff, wbytes);
 	}
 
-	uint16_t sum = _sumBytes(buf, n_data + 2);
-	prev_sum += sum;
-	buf[n_data + 2] = sum & checksum_mask;
-	//printf("\nmessage:\n");
-	//for (size_t i=0;i<n_data+3;i++) {
-	//printf("%d\t", buf[i]);
-	//}
-	//printf("\n");
-	return write(_uart, buf, n_data + 3);
-}
+	wbytes = wbytes + (send_checksum ? 4 : 2);
 
-int roboclawTest(const char *deviceName, uint8_t address,
-		 uint16_t pulsesPerRev)
-{
-	printf("roboclaw test: starting\n");
-
-	// setup
-	RoboClaw roboclaw(deviceName, address, pulsesPerRev);
-	roboclaw.setMotorDutyCycle(RoboClaw::MOTOR_1, 0.3);
-	roboclaw.setMotorDutyCycle(RoboClaw::MOTOR_2, 0.3);
-	char buf[200];
-
-	for (int i = 0; i < 10; i++) {
-		usleep(100000);
-		roboclaw.readEncoder(RoboClaw::MOTOR_1);
-		roboclaw.readEncoder(RoboClaw::MOTOR_2);
-		roboclaw.printStatus(buf, 200);
-		printf("%s", buf);
+	if (send_checksum) {
+		uint16_t sum = _calcCRC(buf, wbytes - 2);
+		buf[wbytes - 2] = (sum >> 8) & 0xFF;
+		buf[wbytes - 1] = sum & 0xFFu;
 	}
 
-	roboclaw.setMotorDutyCycle(RoboClaw::MOTOR_1, -0.3);
-	roboclaw.setMotorDutyCycle(RoboClaw::MOTOR_2, -0.3);
+	int count = write(_uart, buf, wbytes);
 
-	for (int i = 0; i < 10; i++) {
-		usleep(100000);
-		roboclaw.readEncoder(RoboClaw::MOTOR_1);
-		roboclaw.readEncoder(RoboClaw::MOTOR_2);
-		roboclaw.printStatus(buf, 200);
-		printf("%s", buf);
+	if (count < (int) wbytes) { // Did not successfully send all bytes.
+		PX4_ERR("Only wrote %d out of %d bytes", count, (int) wbytes);
+		return -1;
 	}
 
-	printf("Test complete\n");
-	return 0;
-}
+	// READ
+
+	FD_ZERO(&_uart_set);
+	FD_SET(_uart, &_uart_set);
 
-// vi:noet:smarttab:autoindent:ts=4:sw=4:tw=78
+	uint8_t *rbuff_curr = rbuff;
+	size_t bytes_read = 0;
+
+	// select(...) returns as soon as even 1 byte is available. read(...) returns immediately, no matter how many
+	// bytes are available. I need to keep reading until I get the number of bytes I expect.
+	while (bytes_read < rbytes) {
+		err_code = select(_uart + 1, &_uart_set, nullptr, nullptr, &_uart_timeout);
```

Comment:
If we ever drop a byte on the wire, we would then loop forever here, right?

I think usually this is done using some kind of timeout, and if we hit that we say it was not successful and return.